jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/Request.java

```diff
@@ -1033,6 +1033,27 @@ static <T> T as(Request request, Class<T> type)
         return null;
     }
 
+    /** Unwrap a Request back to the given type, ensuring that we do not cross a
+     * context boundary (as might be the case during cross-context dispatch).
+     * @param request the possibly wrapped request to unwrap
+     * @param type the type to unwrap to
+     * @param context the limiting context
+     * @return the request unwrapped back to the given type, or null if it cannot be
+     * unwrapped to that type or a context boundary is crossed.
+     */
+    static <T extends Request> T as(Request request, Class<T> type, Context context)
+    {
+        while (request != null)
+        {
+            if (type.isInstance(request))
+                return (T)request;
+            request = request instanceof Request.Wrapper wrapper ? wrapper.getWrapped() : null;
+            if (request == null || request.getContext() != context)
+                return null;
+        }
+        return null;
```

Comment:
I've change my mind on the order of the test, as it will avoid the double null check:

```suggestion
        while (request != null)
        {
            if (request.getContext() != context)
                return null;
            if (type.isInstance(request))
                return (T)request;
            request = request instanceof Request.Wrapper wrapper ? wrapper.getWrapped() : null;
        }
        return null;
```